File: strategy/java/ShoppingCart.java
package Strategy;

import java.util.ArrayList;
import java.util.List;

// UML diyagramındaki Context sınıfına denk gelmektedir.
// UML diyagramındaki Strategy(IPayment) arayüzüne ait referansı tutmaktadır.
class ShoppingCart {
    private List<Item> items;
    private IPayment payment;

    public ShoppingCart() {
        items = new ArrayList<Item>();
    }

    public void setPaymentMethod(IPayment payment) {
        this.payment = payment;
    }

    public void addItem(Item item) {
        this.items.add(item);
    }

    public int calculateTotal() {
        int sum = 0;
        for (Item item : this.items) {
            sum += item.getPrice();
        }
        return sum;
    }

    public void Pay() {
        int amount = calculateTotal();
        this.payment.pay(amount);
    }
}

File: strategy/kotlin/ShoppingCart.kt
package Strategy

import java.util.ArrayList

// UML diyagramındaki Context sınıfına denk gelmektedir.
// UML diyagramındaki Strategy(IPayment) arayüzüne ait referansı tutmaktadır.
internal class ShoppingCart {
    private val items: MutableList<Item>
    private var payment: IPayment? = null

    init {
        items = ArrayList()
    }

    fun setPaymentMethod(payment: IPayment) {
        this.payment = payment
    }

    fun addItem(item: Item) {
        this.items.add(item)
    }

    fun calculateTotal(): Int {
        var sum = 0
        for (item in this.items) {
            sum += item.price
        }
        return sum
    }

    fun Pay() {
        val amount = calculateTotal()
        this.payment!!.pay(amount)
    }
}